src/az_ulib_ipc/az_ulib_ipc.c

```diff
@@ -0,0 +1,352 @@
+#include "ulib_config.h"
+#include "ulib_port.h"
+#include "ulib_result.h"
+#include "ulog.h"
+
+/*
+ * IPC is a singleton component, and shall be initialized only once.
+ */
+static _az_ulib_ipc *ipc = NULL;
+
+static _az_ulib_ipc_interface *get_interface_by_name_and_version(
+    const char *const name,
+    az_ulib_version version,
+    AZ_ULIB_VERSION_MATCH_CRITERIA match_criteria) {
+  _az_ulib_ipc_interface *result = NULL;
+
+  for (size_t i = 0; i < AZ_ULIB_CONFIG_MAX_IPC_INTERFACE; i++) {
+    if ((ipc->interface_list[i].interface_descriptor != NULL)
+        && (strcmp(ipc->interface_list[i].interface_descriptor->name, name) == 0)
+        && az_ulib_version_match(
+               ipc->interface_list[i].interface_descriptor->version, version, match_criteria)) {
+      result = &(ipc->interface_list[i]);
+      break;
+    }
+  }
+
+  return result;
+}
+
+static _az_ulib_ipc_interface *get_first_free() {
+  _az_ulib_ipc_interface *result = NULL;
+
+  for (size_t i = 0; i < AZ_ULIB_CONFIG_MAX_IPC_INTERFACE; i++) {
+    if ((ipc->interface_list[i].interface_descriptor == NULL)
+        && (ipc->interface_list[i].ref_count == 0)) {
+      result = &(ipc->interface_list[i]);
+      break;
+    }
+  }
+
+  return result;
+}
+
+static AZ_ULIB_RESULT get_instance(_az_ulib_ipc_interface *ipc_interface) {
+  AZ_ULIB_RESULT result;
+  if (ipc_interface->ref_count >= AZ_ULIB_CONFIG_MAX_IPC_INSTANCES) {
+    result = AZ_ULIB_BUSY_ERROR;
+  } else {
+    result = AZ_ULIB_SUCCESS;
+    AZ_ULIB_PORT_ATOMIC_INC_W(&(ipc_interface->ref_count));
+  }
+  return result;
+}
+
+AZ_ULIB_RESULT _az_ulib_ipc_init_no_test(_az_ulib_ipc *handle) {
+  /*az_ulib_ipc_init_succeed*/
+  ipc = handle;
+
+  az_pal_os_lock_init(&(ipc->lock));
+
+  for (size_t i = 0; i < AZ_ULIB_CONFIG_MAX_IPC_INTERFACE; i++) {
+    ipc->interface_list[i].ref_count = 0;
+    ipc->interface_list[i].running_count = 0;
+    ipc->interface_list[i].interface_descriptor = NULL;
+  }
+
+  return AZ_ULIB_SUCCESS;
+}
+
+AZ_ULIB_RESULT _az_ulib_ipc_init(_az_ulib_ipc *handle) {
+  AZ_UCONTRACT(
+      /*az_ulib_ipc_init_double_initialization_failed*/
+      AZ_UCONTRACT_REQUIRE_NULL(ipc, AZ_ULIB_ALREADY_INITIALIZED_ERROR),
+      /*az_ulib_ipc_init_with_null_handle_failed*/
+      AZ_UCONTRACT_REQUIRE_NOT_NULL(handle, AZ_ULIB_ILLEGAL_ARGUMENT_ERROR));
+  return _az_ulib_ipc_init_no_test(handle);
+}
+
+AZ_ULIB_RESULT _az_ulib_ipc_deinit_no_test(void) {
+  AZ_ULIB_RESULT result;
+
+  result = AZ_ULIB_SUCCESS;
+  for (size_t i = 0; i < AZ_ULIB_CONFIG_MAX_IPC_INTERFACE; i++) {
+    if ((ipc->interface_list[i].interface_descriptor != NULL)
+        || (ipc->interface_list[i].ref_count != 0) || (ipc->interface_list[i].running_count != 0)) {
+      /*az_ulib_ipc_deinit_with_published_interface_failed*/
+      /*az_ulib_ipc_deinit_with_instace_failed*/
+      result = AZ_ULIB_BUSY_ERROR;
+      break;
+    }
+  }
+
+  if (result == AZ_ULIB_SUCCESS) {
+    /*az_ulib_ipc_deinit_succeed*/
+    ipc = NULL;
+  }
```

Comment:
Can't we have race conditions between init/deinit? As well publish/deinit?

Reply:
Both init and deinit are not thread safe, the "lock" is created on the init, and destroyed in the deinit, so there is no way to protect this 2 functions.

---
In reply to: [337795488](https://github.com/Azure/azure-ulib-c/pull/40#discussion_r337795488) [](ancestors = 337795488)